### PR TITLE
rumqttc: resume session only if CONNACK with session present 1

### PR DESCRIPTION
**Pull request by xiaocq2001**
## Type of change

Implement for https://github.com/bytebeamio/rumqtt/issues/863.

Bug fix (non-breaking change which fixes an issue):

In MQTT, session is not always resumed, but in current mqttc the pending publishes are restored anyway.
With this fix the CONNACK session present returned from broker is checked, to confirm if a session is actually resumed, if not the pending publishes are cleaned and new session is started.

## Checklist:

- [ ] Formatted with `cargo fmt`
- [ ] Make an entry to `CHANGELOG.md` if it's relevant to the users of the library. If it's not relevant mention why.

### CODE REVIEW

File: rumqttc/src/v5/eventloop.rs [hunk 1]
```diff
@@ -406,12 +407,22 @@ async fn mqtt_connect(
...
             if let Some(props) = &connack.properties {
+                // Override local keep_alive value if set by server.
                 if let Some(keep_alive) = props.server_keep_alive {
                     options.keep_alive = Duration::from_secs(keep_alive as u64);
                 }
                 network.set_max_outgoing_size(props.max_packet_size);
+                // Override local session_expiry_interval value if set by server.
+                if (props.session_expiry_interval).is_some() {
+                    options.set_session_expiry_interval(props.session_expiry_interval);
+                }
```
Comment (de-sh) [inline rumqttc/src/v5/eventloop.rs:425]: These changes should rather go into #854

Comment (xiaocq2001) [reply rumqttc/src/v5/eventloop.rs:425]: Fine, I'll move that.

**Review by de-sh — CHANGES_REQUESTED** [review]
https://github.com/xiaocq2001/rumqtt/pull/1

**Comment by swanandx** [comment]
can you please fix the code / tests? 

the `reconnection_resends_unacked_packets_from_the_previous_connection_first` test is failing, i think we need to connect with clean session false so that broker will save the state right?

**Comment by xiaocq2001** [comment]
To test, I think we need https://github.com/bytebeamio/rumqtt/pull/854. With that, we need to connect with clean session false and session expiry interval non-zero to let broker save state.

**Comment by xiaocq2001** [comment]
> To test, I think we need #854. With that, we need to connect with clean session false and session expiry interval non-zero to let broker save state.

Facing an issue on connection. Due to MQTT protocol difference between versions, in v5 session_expiry_interval is supported but in v3 it is not, what is the suggestion on handling such difference in tests?

**Comment by de-sh** [comment]
> > To test, I think we need #854. With that, we need to connect with clean session false and session expiry interval non-zero to let broker save state.
> 
> Facing an issue on connection.
> 
> ```
> ---- reconnection_resumes_from_the_previous_state stdout ----
> Polled = Ok(Incoming(ConnAck(ConnAck { session_present: false, code: Success })))
> Polled = Ok(Outgoing(Publish(1)))
> Polled = Ok(Incoming(PubAck(PubAck { pkid: 1 })))
> Polled = Ok(Outgoing(Publish(2)))
> Polled = Ok(Incoming(PubAck(PubAck { pkid: 2 })))
> Polled = Ok(Outgoing(Publish(3)))
> Polled = Ok(Outgoing(Publish(4)))
> Polled = Ok(Outgoing(Publish(5)))
> Polled = Ok(Outgoing(PingReq))
> Polled = Ok(Outgoing(Publish(6)))
> Polled = Ok(Outgoing(Publish(7)))
> Polled = Ok(Outgoing(Publish(8)))
> Polled = Ok(Outgoing(Publish(9)))
> Polled = Ok(Outgoing(Publish(10)))
> Polled = Err(MqttState(AwaitPingResp))
> Polled = Ok(Incoming(ConnAck(ConnAck { session_present: false, code: Success })))
> ```
> 
> It seems on host side session is not stored?
> 
> Another question: due to MQTT protocol difference between versions, in v5 session_expiry_interval is supported but in v3 it is not, what is the suggestion on handling such difference in tests?

https://github.com/xiaocq2001/rumqtt/pulls solves this inconsistency in the tests

**Comment by xiaocq2001** [comment]
I realize that in failing tests, the broker is started by `Broker::new`, which starts a broker without any stored state, that's why broker send ConnAck with session present false, we need some new method to introduce disconnection, maybe like what is in async_manual_ack example.

**Review by de-sh — APPROVED** [review]

**Pull request merged**